Set up CI build

penguinland opened the pull request:
Naturally I haven't been able to run this in a Github action runner yet, but `make setup`, `make lint`, and `make archive.tar.gz` all succeed on my local machine.

A doubt: in `setup.sh`, I might run `sudo apt-get install clang-format`. Within the CI build, will it get stuck waiting to receive a password? We'll find out...

penguinland committed:
update to the latest SDK version

penguinland committed:
nit: remove trailing whitespace, fix typo

penguinland committed:
no idea where this came from; conan might have added it on its own

penguinland committed:
add a make target for setup

penguinland committed:
I didn't touch this, I think conan changed it again

penguinland committed:
add module.tar.gz makefile target

penguinland committed:
add make target for linter, add linter installation to setup

penguinland committed:
clean up after setup is over

penguinland committed:
add CI lint check

penguinland commented on CMakeUserPresets.json:
```diff
@@ -4,6 +4,7 @@
         "conan": {}
     },
     "include": [
+        "build-conan/build/Release/generators/CMakePresets.json",
```
I didn't add this line; I think `conan` did. :shrug:

lia-viam replied:
so `CMakeUserPresets.json` should generally not be checked into CI and could be added to a `.gitignore`, see https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html#id3

penguinland commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,24 @@
+      - uses: actions/checkout@v4
+
+      - name: Install dependencies
+        run: make setup
```
This needs better caching, or else every time you push another commit to your PR, it'll rebuild protobuf (and the SDK) from source. Looking into how to do that now. In the meantime, we can either merge this with a TODO to fix this soon, or else figure out how to fix it now.

penguinland commented:
...and the checks aren't running because I didn't make a PR into `main`. I'm gonna relax those constraints, planning to revert that change later...

penguinland committed:
REVERTME: run the checks on any push or any PR

penguinland committed:
cache the conan artifacts from 'make setup'

bhaney reviewed (approved):
LGTM! Will be exciting to get the C++ version published to registry

bhaney commented:
Actually - could you delete the "packaging/" directory, as well as the part of the makefile that makes the AppImage? Also the .artifact directory, which was used by the Go code. Those are no longer necessary

penguinland committed:
pr feedback: delete packaging/ .artifact/ and appimage

penguinland committed:
clarify more in the makefile

penguinland committed:
don't bother copying binaries around when zipping the module

penguinland committed:
allow gzip to write new versions of the .tar.gz over the old one

penguinland committed:
bugfix: cache ~/.conan2, not ~/.conan

penguinland commented:
Done. I've also moved things around slightly in the makefile: we no longer copy the executable elsewhere when making module.tar.gz, and I've updated meta.json to reflect this updated location.

penguinland committed:
simplify zipping the module

penguinland commented:
I've made one more change, to simplify creating the .tar.gz into a one-liner instead of a two-liner. I haven't pushed to github yet because I'd like that commit to demonstrate that the CI caching is working correctly, which means waiting for the current run to complete.

penguinland commented:
Yes, the cache is working! Now to revert the thing about running the checks on non-`main` branches, ~and also add cache use to the deploy action...~ _edit: it's already in there!_

penguinland committed:
Revert "REVERTME: run the checks on any push or any PR"

This reverts commit 033cc9e3f696c4a7ef25cc870e361ad1bbb8227f.

penguinland commented:
...and I'm waiting for the checks to run, but they never will because this is not a PR into `main` and I reverted the commit that runs them everywhere! :sweat_smile:

lia-viam commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,31 @@
+
+      - uses: actions/cache@v3
+        with:
+          path: ~/.conan2
```
Looks like here it's spelled `.conan2` which I believe is correct but above it says `.conan`

penguinland replied:
:facepalm: This is totally right! I wish I'd caught that earlier, thanks

lia-viam commented on bin/setup.sh:
```diff
@@ -0,0 +1,39 @@
+conan install . --output-folder=build-conan --build=missing
+cmake . --preset=conan-release
+cmake --build --preset=conan-release -j 8
+conan create . -o "&:shared=False" --build=missing
```
It looks like there is some redundant building happening here. If I'm following correctly the intention is to create a `viam-cpp-sdk` conan package which can then be consumed as a dependency by the `tflite` module. If so, only this final line with `conan create .` is necessary. The steps being performed are

- 19: Use `conan` to obtain the dependencies of the C++ SDK
- 20: Generate a CMake build for building the SDK from source using the dependencies generated in `19`
- 21: Build the SDK from source using the files generated in `20`
- 22: Build the SDK conan package from source; will reuse the dependencies obtained in `19` but none of the stuff from `20` & `21` which are then deleted in `26`

Once you do line `22` the SDK package exists in the local conan package cache (in `~/.conan2`) and will be found by local builds which declare `viam-cpp-sdk` as a dependency

lia-viam reviewed (commented):
hope this is helpful, please don't hesitate to @ me in any subsequent PRs involving conan stuff!
